chore(yellow_paper): public<>private messaging (#3491)
AztecProtocol/aztec-packages: yellow-paper/docs/calls/public_private_messaging.md

This way, you can destroy your private state and create them in public within the same transaction or indirectly assert constraints on the execution of the private functions with latest data.

### Handling Privacy Leakage and `msg.sender`
In the above design, the sequencer only sees the public part of the call stack along with any new commitments, nullifiers etc that were created in the private transaction i.e. should learns nothing more of the private transaction (such as its origin, execution logic etc).

But what if the enqueued public function makes use of `msg_sender` which is meant to use 

Specifically, when the call stack is passed to the kernel circuit, the kernel should assert the `msg_sender` is 0 and hash appropriately. `msg_sender` could be the contract address too instead of `0`, but it leaks which contract is calling the public method and therefore leaks which contract the user was interacting with in private land.

### Reverts

> If the private part of the transaction reverts, then public calls are never enqueued. But if the public part of the transaction reverts, it should still revert the entire transaction i.e. the sequencer should drop the execution results of the private part of the transaction and not include those in the state transitioner smart contract. However, since the sequencer had to execute your transaction, appropriate fee will be charged. Reverting in public causing the whole transaction to be dropped enables existing paradigms of ethereum where your valid transaction can revert because of altered state e.g., trade incurring too much slippage.

Reviewer: It should drop the state changes (effects of the tx), guess that is the same as the execution result, but for result I'm often thinking the return values etc.

Also, we are not including it in the contract, but in the block input to the contract. 

A note here on reverts, I still think the first nullifier is desired to ensure that the tx cannot just be to drain fees. So there is still some effect of it, kinda like the nonce increasing 🤷.

## Public -> Private Messaging

> Since public functions execute after private functions, it isn't possible for public to call a private function in the same transaction. Nevertheless, it is quite useful for public functions to have a message passing system to private. A public function could add messages to an append only merkle tree to save messages from a public function call, that can later be executed by a private function. Note, only a transaction coming after the one including the message from a public function can consume it. In practice this means that unless you are the sequencer it will not be within the same rollup.

Reviewer: I think it is nicer to consistently use `consume` than having both `consume` and `executed`. 

```suggestion
Since public functions execute after private functions, it isn't possible for public to call a private function in the same transaction. Nevertheless, it is quite useful for public functions to have a message passing system to private. A public function could add messages to an append only merkle tree to save messages from a public function call, that can later be consumed by a private function. Note, only a transaction coming after the one including the message from a public function can consume it. In practice this means that unless you are the sequencer or work with it, this will not be within the same rollup.
```

> To elaborate, a public function may not have read access to encrypted private state in the note hash tree, but it can write to it. You could create a note in the public domain, compute it's note hash which gets passed to the inputs of the public VM which adds the hash to the note hash tree. The user who wants to redeem the note can add the note preimage to their PXE and then redeem/nullify the note in the private domain at a later time. 

Reviewer: You could read, but that would rely on it being public so might just be a fine note to have 🤷.

```suggestion
To elaborate, a public function should not read encrypted private state from the note hash tree, since the information is then made public, but it could be made to write notes to it.
You could create a note in the public domain, compute it's note hash which gets passed to the inputs of the public VM which adds the hash to the note hash tree. The user who wants to redeem the note can add the note preimage to their PXE and then redeem/nullify the note in the private domain at a later time.
```
